libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/ApplicationBuilderExtensions.cs

```diff
@@ -86,10 +86,55 @@ public static IApplicationBuilder UseNamedPipes(this IApplicationBuilder applica
                 throw new ArgumentNullException(nameof(applicationBuilder));
             }
 
-            var bot = applicationBuilder.ApplicationServices.GetService(typeof(IBot)) as IBot;
-            _ = (applicationBuilder.ApplicationServices.GetService(typeof(IBotFrameworkHttpAdapter)) as BotFrameworkHttpAdapter).ConnectNamedPipeAsync(pipeName, bot, audience);
+            return applicationBuilder.UseNamedPipes(pipeName, audience, null, null);
+        }
+
+        /// <summary>
+        /// Enables named pipes for this application.
+        /// </summary>
+        /// <param name="applicationBuilder">The application builder that defines the bot's pipeline.<see cref="IApplicationBuilder"/>.</param>
+        /// <param name="pipeName">The name of the named pipe to use when creating the server.</param>
+        /// <param name="audience">The specified recipient of all outgoing activities.</param>
+        /// <param name="appId">The bot's application id.</param>
+        /// <param name="callerId">The caller id.</param>
+        /// <returns>A reference to this instance after the operation has completed.</returns>
+        public static IApplicationBuilder UseNamedPipes(this IApplicationBuilder applicationBuilder, string pipeName, string audience, string appId, string callerId)
+        {
+            if (applicationBuilder == null)
+            {
+                throw new ArgumentNullException(nameof(applicationBuilder));
+            }
+
+            var bot = applicationBuilder.ApplicationServices.GetRequiredService<IBot>();
+
+            var adapter = applicationBuilder.ApplicationServices.GetRequiredService<IBotFrameworkHttpAdapter>();
+
+            if (typeof(BotFrameworkHttpAdapter).IsAssignableFrom(adapter.GetType()))
+            {
+                // back compatibility: support for the BotFrameworkHttpAdapter
+                _ = ((BotFrameworkHttpAdapter)adapter).ConnectNamedPipeAsync(pipeName, bot, audience);
+            }
+            else if (typeof(CloudAdapter).IsAssignableFrom(adapter.GetType()))
+            {
+                // if any of these values are null then attempt to pull them from configuration
+                var configuration = applicationBuilder.ApplicationServices.GetService<IConfiguration>();
+                if (configuration != null)
+                {
+                    audience = audience ?? configuration.GetSection("ToChannelFromBotOAuthScope")?.Value ?? GetBuiltinDefaultAudience(configuration);
```

Comment:
So I tried to keep the logic in the CloudAdapter class itself clean and so pushed configuration and defaulting logic to this helper.

Some of the complexity comes from not breaking the current method. It would be easy to get into an ambiguous overload. So the new overload just requires everything.

Yes surprisingly complicated - despite the appropriate (and improved) use of the DI helpers.

More fundamentally starting the pipe from this extension seems awkward - and dropping the return task. But that is what the current method does. (Ultimately the protocol semantics are application logic to application logic so there isn't much that can be done at this level anyhow.)

Anyhow this is just a helper - for example the adaptive runtime does not need to use this helper at all.